In [45]:
import pandas as pd
import numpy as np

## Обзор данных

Для считывания данных используется функция `pd.read_csv`. У этой функции много аргументов (см. документацию), первый из которых - название считываемого файла с данными, второй - sep - разделитель, по умолчанию это запятая. Данные в csv файлах (обычно, но не всегда) записаны в виде "таблицы". Одна запись - одна строка, данные столбцов разделяются запятыми.

Для работы с файлом его, как и обычный текстовый файл (впрочем, это и есть обычный текстовый файл!), удобнее положить в ту же директорию, где лежит ноутбук. Затем вы можете прямо через jupyter открыть файл `csv` и увидеть, как он выглядит.

In [46]:
f = pd.read_csv("telecom_churn.csv")

Посмотреть начало и конец таблицы (по умолчанию 5 строк, но это настраиваемый параметр) - функции `head` и `tail`. Описание данных - `describe`, `info`. Размер таблицы - `shape`, названия колонок - `columns`. Обратите внимание, что `columns` возвращает итерируемый объект.

In [16]:
f.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [17]:
f.head(7)

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   
5    AL             118        510                Yes              No   
6    MA             121        510                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  
5                       0  False  
6                       3  False

In [18]:
f.tail()

State  Account length  Area code International plan Voice mail plan  \
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
3328                       2  False  
3329                       3  False  
3330                       2  False  
3331                       2  False  
3332                       0  False

In [19]:
f.tail(7)

State  Account length  Area code International plan Voice mail plan  \
3326    OH              96        415                 No              No   
3327    SC              79        415                 No              No   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3326                      0              106.6              128   
3327                      0              134.7               98   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3326             18.12              284.8               87             24.21   
3327             22.90              189.7               68             16.12   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
3326                178.9                 92                8.05   
3327                221.4                128                9.96   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3326                14.9                 7               4.02   
3327                11.8                 5               3.19   
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
3326                       1  False  
3327                       2  False  
3328                       2  False  
3329                       3  False  
3330                       2  False  
3331                       2  False  
3332                       0  False

In [20]:
f.tail(100)

State  Account length  Area code International plan Voice mail plan  \
3233    OK             112        415                 No              No   
3234    DE              75        510                 No             Yes   
3235    AZ              97        408                 No             Yes   
3236    AK             121        408                 No             Yes   
3237    MI             142        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3233                      0              166.0               79   
3234                     28              200.6               96   
3235                     25              141.0              101   
3236                     34              245.0               95   
3237                      0              140.8              140   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3233             28.22               74.6              100              6.34   
3234             34.10              164.1              111             13.95   
3235             23.97              212.0               85             18.02   
3236             41.65              216.9               66             18.44   
3237             23.94              228.6              119             19.43   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
3233                247.9                 74               11.16   
3234                169.6                153                7.63   
3235                175.2                138                7.88   
3236                112.4                125                5.06   
3237                152.9                 88                6.88   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3233                 6.3                 7               1.70   
3234                 2.5                 5               0.68   
3235                 4.9                 2               1.32   
3236                 7.5                 8               2.03   
3237                10.9                 7               2.94   
...                  ...               ...                ...   
3328                 9.9                 6            

In [21]:
f.shape

(3333, 20)

In [22]:
f.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [ ]:
f.describe()

In [ ]:
f.info()

Небольшое, но важное отступление. Глубоко понимать это пока не обязательно. Обратите внимание, что функция `print` возвращает строковое представление объекта и она не необходима, чтобы вывести объект. 2 ячейки сработают одинаково:

In [23]:
a = 1
print(a)

1


In [24]:
a = 1
a

1

При этом для некоторых объектов (для которых стандартные функции `__str__` и `__repr__` возвращают разные значения) два способа вывода сработают несколько по-разному:

In [25]:
a = np.array([1.0, 2.0])
print(a)

[1. 2.]


In [26]:
a = np.array([1.0, 2.0])
a

array([1., 2.])

Такая же ситуация с pandas dataframe'ами. Без print получится красивее, чем с ним, но фактически это одно и то же:

In [27]:
print(f)

     State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                   

In [28]:
f

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

2 варианта синтаксиса для просмотра содержимого столбца:

In [29]:
f.State

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

In [30]:
f['State']

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

Обратите внимание, что таблица и столбцы имеют разный тип:

In [31]:
type(f), type(f.State), type(f['State'])

(pandas.core.frame.DataFrame,
 pandas.core.series.Series,
 pandas.core.series.Series)

Тем не менее, мы можем сделать именно датафрейм из одного столбца следующим образом:

In [32]:
f[['State']]

State
0       KS
1       OH
2       NJ
3       OH
4       OK
...    ...
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN

[3333 rows x 1 columns]

In [33]:
type(f[['State']]), type(f['State'])

(pandas.core.frame.DataFrame, pandas.core.series.Series)

Это способ обращения к "срезам" таблицы по названиям столбцов. Еще пример:

In [34]:
f[['State', 'Churn']]

State  Churn
0       KS  False
1       OH  False
2       NJ  False
3       OH  False
4       OK  False
...    ...    ...
3328    AZ  False
3329    WV  False
3330    RI  False
3331    CT  False
3332    TN  False

[3333 rows x 2 columns]

Просмотр уникальных значений в столбце и их числа:

In [ ]:
f.State.unique()

In [ ]:
f.State.nunique()

Обращение к элементу ообъекта `pd.Series`:

In [ ]:
f.State[0]

Полный набор функций, применимых к столбцу можно посмотреть следующим образом. Здесь используется стандартная функция `__dir__()`, которая выводит список атрибутов и методов объекта.

In [ ]:
f.State.__dir__()

С `pd.DataFrame` привычный способ обращения по индексам не сработает:

In [ ]:
f[0][0]

Основной способ - использование `loc` и `iloc`. `loc` дает индексацию по "настоящим" индексам - названиям столбцов и названиям строк. В нашем случае, названия строк - целые числа, как обычные индексы. `iloc` позволяет осуществлять стандартную численную индексацию, как для `numpy` массивов.

In [ ]:
f.iloc[0] # первая строка

In [ ]:
f.iloc[0][0]

In [ ]:
f.iloc[0, 0]

In [ ]:
f.iloc[[0, 0]]

In [ ]:
f.iloc[[1, 2, 5, 1]]

In [ ]:
f.loc[[1, 2, 5, 1]]

In [ ]:
f.loc[[1, 2, 5, 1], ["State", "Account length"]]

In [ ]:
f.iloc[[1, 2, 5, 1], [0, 1]]

Следующая ячейка вернет ошибку:

In [ ]:
f.iloc[[1, 2, 5, 1], ["State", "Account length"]]

In [ ]:
f.iloc[:5, :4]

In [ ]:
f.loc[:, ["State", "Account length"]]

In [ ]:
f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

Обратите внимание, что `loc` включает и строки соответствующие и начальному, и конечному индексам.

Сохраним этот датафрейм в новую переменную и рассмотрим отличие loc и iloc в работе по целочисленным индексам строк

In [ ]:
new_f = f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

In [ ]:
new_f.iloc[:5]

In [ ]:
new_f.loc[:5]

**Упражнение 1.** Объясните различие поведения `loc` и `iloc` для `new_f`.

#### Создание датафрейма и присваивание по индексам

In [5]:
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]}, index = ["first", "second", "third"])
df

A  B
first   1  4
second  2  5
third   3  6

In [6]:
df.iloc[0, 0] = 100
df

A  B
first   100  4
second    2  5
third     3  6

In [7]:
df.loc["second", "B"] = 31
df

A   B
first   100   4
second    2  31
third     3   6

In [8]:
df.iloc[2] = 5
df

A   B
first   100   4
second    2  31
third     5   5

In [9]:
df.loc["B"] = 0
df

A   B
first   100   4
second    2  31
third     5   5
B         0   0

In [10]:
df.loc[:, "B"] = 0
df

A  B
first   100  0
second    2  0
third     5  0
B         0  0

In [11]:
df = pd.DataFrame({"A": [1, 2, 3, 4], "B": [5, 6, 7, 8]})
df

A  B
0  1  5
1  2  6
2  3  7
3  4  8

In [12]:
df.iloc[1:2, :] = -1
df

A  B
0  1  5
1 -1 -1
2  3  7
3  4  8

In [13]:
df.loc[1:2, :] = -2
df

A  B
0  1  5
1 -2 -2
2 -2 -2
3  4  8

Еще раз убеждаемся, что `loc` включает оба конца "среза", а `iloc` - полуинтервал, как например обычные срезы.

#### Создание нового столбца

In [ ]:
df["C"] = 1
df

In [ ]:
df["D"] = df["A"] + df["B"]
df

In [ ]:
df["E"] = df["D"].apply(lambda x: x ** 2)
df

In [ ]:
df["F"] = df[["A", "C", "E"]].apply(lambda x: x.sum(), axis=1)
df

**Упражнение 2.** Объясните последний результат.

## Срезы с условиями и аггрегация

In [35]:
f[f.Churn == True]

State  Account length  Area code International plan Voice mail plan  \
10      IN              65        415                 No              No   
15      NY             161        415                 No              No   
21      CO              77        408                 No              No   
33      AZ              12        408                 No              No   
41      MD             135        408                Yes             Yes   
...    ...             ...        ...                ...             ...   
3301    CA              84        415                 No              No   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   
3322    MD              62        408                 No              No   
3323    IN             117        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
10                        0              129.1              137   
15                        0              332.9               67   
21                        0               62.4               89   
33                        0              249.6              118   
41                       41              173.1               85   
...                     ...                ...              ...   
3301                      0              280.0              113   
3304                      0              186.1              114   
3320                      0              140.0              101   
3322                      0              321.1              105   
3323                      0              118.4              126   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10               21.95              228.5               83             19.42   
15               56.59              317.8               97             27.01   
21               10.61              169.9              121             14.44   
33               42.43              252.4              119             21.45   
41               29.43              203.9              107             17.33   
...                ...                ...              ...               ...   
3301             47.60              202.2               90             17.19   
3304             31.64              198.6              140             16.88   
3320             23.80              196.4               77             16.69   
3322             54.59              265.5              122             22.57   
3323             20.13              249.3               97             21.19   

      Total night minutes  Total night calls  Total night charge  \
10                  208.8                111                9.40   
15                  160.6                128                7.23   
21                  209.6                 64                9.43   
33                  280.2                 90               12.61   
41                  122.2                 78                5.50   
...                   ...                ...                 ...   
3301                156.8                103                7.06   
3304                206.5                 80                9.29   
3320                120.1                133                5.40   
3322                180.5                 72                8.12   
3323                227.0                 56               10.22   

      Total intl minutes  Total intl calls  Total intl charge  \
10                  12.7                 6               3.43   
15                   5.4                 9               1.46   
21                   5.7                 6               1.54   
33                  11.8                 3               3.19   
41                  14.6                15               3.94   
...                  ...               ...                ...   
3301                10.4                 4            

In [36]:
f[f["Area code"] == 408]

State  Account length  Area code International plan Voice mail plan  \
3       OH              84        408                Yes              No   
8       LA             117        408                 No              No   
12      IA             168        408                 No              No   
16      ID              85        408                 No             Yes   
21      CO              77        408                 No              No   
...    ...             ...        ...                ...             ...   
3309    VT             100        408                Yes              No   
3312    SC             181        408                 No              No   
3313    ID             127        408                 No              No   
3322    MD              62        408                 No              No   
3325    OH              78        408                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
3                         0              299.4               71   
8                         0              184.5               97   
12                        0              128.8               96   
16                       27              196.4              139   
21                        0               62.4               89   
...                     ...                ...              ...   
3309                      0              219.4              112   
3312                      0              229.9              130   
3313                      0              102.8              128   
3322                      0              321.1              105   
3325                      0              193.4               99   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3                50.90               61.9               88              5.26   
8                31.37              351.6               80             29.89   
12               21.90              104.9               71              8.92   
16               33.39              280.9               90             23.88   
21               10.61              169.9              121             14.44   
...                ...                ...              ...               ...   
3309             37.30              225.7              102             19.18   
3312             39.08              144.4               93             12.27   
3313             17.48              143.7               95             12.21   
3322             54.59              265.5              122             22.57   
3325             32.88              116.9               88              9.94   

      Total night minutes  Total night calls  Total night charge  \
3                   196.9                 89                8.86   
8                   215.8                 90                9.71   
12                  141.1                128                6.35   
16                   89.3                 75                4.02   
21                  209.6                 64                9.43   
...                   ...                ...                 ...   
3309                255.3                 95               11.49   
3312                262.4                110               11.81   
3313                191.4                 97                8.61   
3322                180.5                 72                8.12   
3325                243.3                109               10.95   

      Total intl minutes  Total intl calls  Total intl charge  \
3                    6.6                 7               1.78   
8                    8.7                 4               2.35   
12                  11.2                 2               3.02   
16                  13.8                 4               3.73   
21                   5.7                 6               1.54   
...                  ...               ...                ...   
3309                12.0                 4            

In [37]:
f[f["Area code"] == 408].loc[3]

State                        OH
Account length               84
Area code                   408
International plan          Yes
Voice mail plan              No
Number vmail messages         0
Total day minutes         299.4
Total day calls              71
Total day charge           50.9
Total eve minutes          61.9
Total eve calls              88
Total eve charge           5.26
Total night minutes       196.9
Total night calls            89
Total night charge         8.86
Total intl minutes          6.6
Total intl calls              7
Total intl charge          1.78
Customer service calls        2
Churn                     False
Name: 3, dtype: object

In [38]:
f[f["Area code"] == 408].loc[3] == f[f["Area code"] == 408].iloc[0]

State                     True
Account length            True
Area code                 True
International plan        True
Voice mail plan           True
Number vmail messages     True
Total day minutes         True
Total day calls           True
Total day charge          True
Total eve minutes         True
Total eve calls           True
Total eve charge          True
Total night minutes       True
Total night calls         True
Total night charge        True
Total intl minutes        True
Total intl calls          True
Total intl charge         True
Customer service calls    True
Churn                     True
Name: 3, dtype: bool

In [39]:
f[f.Churn == True]

State  Account length  Area code International plan Voice mail plan  \
10      IN              65        415                 No              No   
15      NY             161        415                 No              No   
21      CO              77        408                 No              No   
33      AZ              12        408                 No              No   
41      MD             135        408                Yes             Yes   
...    ...             ...        ...                ...             ...   
3301    CA              84        415                 No              No   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   
3322    MD              62        408                 No              No   
3323    IN             117        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
10                        0              129.1              137   
15                        0              332.9               67   
21                        0               62.4               89   
33                        0              249.6              118   
41                       41              173.1               85   
...                     ...                ...              ...   
3301                      0              280.0              113   
3304                      0              186.1              114   
3320                      0              140.0              101   
3322                      0              321.1              105   
3323                      0              118.4              126   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10               21.95              228.5               83             19.42   
15               56.59              317.8               97             27.01   
21               10.61              169.9              121             14.44   
33               42.43              252.4              119             21.45   
41               29.43              203.9              107             17.33   
...                ...                ...              ...               ...   
3301             47.60              202.2               90             17.19   
3304             31.64              198.6              140             16.88   
3320             23.80              196.4               77             16.69   
3322             54.59              265.5              122             22.57   
3323             20.13              249.3               97             21.19   

      Total night minutes  Total night calls  Total night charge  \
10                  208.8                111                9.40   
15                  160.6                128                7.23   
21                  209.6                 64                9.43   
33                  280.2                 90               12.61   
41                  122.2                 78                5.50   
...                   ...                ...                 ...   
3301                156.8                103                7.06   
3304                206.5                 80                9.29   
3320                120.1                133                5.40   
3322                180.5                 72                8.12   
3323                227.0                 56               10.22   

      Total intl minutes  Total intl calls  Total intl charge  \
10                  12.7                 6               3.43   
15                   5.4                 9               1.46   
21                   5.7                 6               1.54   
33                  11.8                 3               3.19   
41                  14.6                15               3.94   
...                  ...               ...                ...   
3301                10.4                 4            

In [40]:
f[f.Churn == True].mean()

/tmp/ipykernel_8033/3514615432.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  f[f.Churn == True].mean()


Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

In [41]:
f[f.Churn == True].sum()

State                     INNYCOAZMDIDWYCOTXDCNYTXINNJLAMSIDMENVMSMSVTOH...
Account length                                                        49587
Area code                                                            211466
International plan        NoNoNoNoYesNoNoNoNoNoNoNoNoNoNoNoNoYesNoNoNoYe...
Voice mail plan           NoNoNoNoYesNoNoYesNoNoNoNoNoNoNoNoNoYesNoNoYes...
Number vmail messages                                                  2471
Total day minutes                                                   99939.5
Total day calls                                                       48945
Total day charge                                                   16989.97
Total eve minutes                                                  102594.1
Total eve calls                                                       48571
Total eve charge                                                    8720.55
Total night minutes                                                 99126.9
Total night 

Выборка по нескольким условиям требует скобок, т.к. операции `&`, `|`, `~` (и, или, не) имеют более высокий приоритет, чем `==`, `>=`, `<=` и арифметические:

In [ ]:
f[(f.Churn == True) & (f["Area code"] == 408)]

Можем посмотреть средние показатели лояльных и нелояльных клиентов (столбец "Churn" показывает, ушел клиент от нас как от оператора связи или нет).

In [43]:
loyal = f[f.Churn == False].mean()
unloyal = f[f.Churn == True].mean()
pd.DataFrame({"loyal": loyal, "unloyal": unloyal})

/tmp/ipykernel_8033/1937728424.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  loyal = f[f.Churn == False].mean()
/tmp/ipykernel_8033/1937728424.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  unloyal = f[f.Churn == True].mean()


loyal     unloyal
Account length          100.793684  102.664596
Area code               437.074737  437.817805
Number vmail messages     8.604561    5.115942
Total day minutes       175.175754  206.914079
Total day calls         100.283158  101.335404
Total day charge         29.780421   35.175921
Total eve minutes       199.043298  212.410145
Total eve calls         100.038596  100.561077
Total eve charge         16.918909   18.054969
Total night minutes     200.133193  205.231677
Total night calls       100.058246  100.399586
Total night charge        9.006074    9.235528
Total intl minutes       10.158877   10.700000
Total intl calls          4.532982    4.163561
Total intl charge         2.743404    2.889545
Customer service calls    1.449825    2.229814
Churn                     0.000000    1.000000

`groupby` позволяет делать группировку по уникальным значениям столбца, которые будут играть роль индекса в полученном датафрейме. При этом также нужна функция агрегации (`mean`, `sum`, `max` и прочие), чтобы отобразить полученный объект.

In [44]:
Name: Name, dtype: objectf.groupby("Churn")

SyntaxError: invalid syntax (1589818681.py, line 1)

In [ ]:
f.groupby("Churn").mean()

Здесь мы сделали агрегацию всех столбцов функцией `mean`. Часто нам нужна агрегация по конкретному или конкретным столбцам. Посмотрим, например, в каком штате больше всего разговаривают по телефону:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"})

Максимальное значение времени:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).max()

Чтобы найти сам штат, можем отсортировать данные:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).sort_values(by="Total day minutes", ascending = False)

**Пример.** Как распределен отток клиентов по штатам? В каких штатах он больше среднего?

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"})

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).loc["LA"]

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]

In [ ]:
m = f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]
new_dataframe = f.groupby(["State"]).agg({"Churn": "mean"})
new_dataframe[new_dataframe.Churn > m]

## Упражнения.

**Упражнение 3.** Найдите среднее количество звонков `Total day calls` для всего датафрейма.

**Упражнение 4.** Найдите среднее количество звонков `Total day calls` для любого выбранного вами штата.

**Упражнение 5.** Создайте датафрейм, в котором будет среднее количество звонков `Total day calls` для каждого штата.

**Упражнение 6.** Оставьте в созданном датафрейме строки только с теми штатами, где количество звонков `Total day calls` больше среднего по исходному датафрейму.

**Упражнение 7.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата.

**Упражнение 8.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата, а также столбец со значениями True и False - ответом на вопрос, больше ли дневных звонков, чем вечерних.

**Упражнение 9.** Найти долю клиентов (отношение их числа к общему количеству клиентов) с `international plan` и `voice mail plan`.

**Упражнение 10.** Найти число уникальных значений `Area code`.

**Упражнение 11.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; число клиентов, звонивших столько раз. Подсказка: используйте функцию агрегации `count`.

**Упражнение 12.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; доля оттока (`Churn`). Построить график.

**Упражнение 13.** Найти среднюю длительность международного (`intl`) звонка.

**Упражнение 14*.** Какие звонки дольше - дневные, вечерние или ночные? Ответ привести в формате DataFrame $3*3$: строки - число минут, число звонков, среднее время звонка. Столбцы - день, вечер, ночь

**Упражнение 15.** Сравнить `Total day charge` для оставшихся и ушедших клиентов.

**Упражнение 16.** Отсортриуйте штаты по `Total day charge` (по возрастанию).

**Упражнение 17.** Сделайте агрегацию по средним показателям для каждой `Area code`.

**Упражнение 18.** Выведите датафрейм размера $3*2$: столбцы State, Churn для 100, 102 и 104 строк исходного датафрейма.

**Упражнение 19.** Создайте датафрейм из 2 столбцов, заполненных произвольными числами. Добавьте третий столбец, равный их сумме квадратов.

**Упражнение 20.** Добавьте к созданному датафрейму четвертый столбец, равный среднему значению первых трех, *используя* функцию `mean`.